### ISSUE TITLE
Intervals/ranges in NodaTime 2.0

### ISSUE THREAD

**Issue by lundmikkel** [nodatime/nodatime#649]
NodaTime has a wonderful set of time and date related data types. However, it generally lacks support for intervals/ranges. NodaTime 1.x only contains `Interval` (a `Instant` interval), whereas NodaTime 2.0 seems to introduce a `DateInterval` as well. Most types of intervals are not supported in the library, e.g. `DateTimeInterval` and `TimeInterval`.

Neither `Interval` nor `DateInterval` support [interval relations](https://en.wikipedia.org/wiki/Allen's_interval_algebra). There are a lot of helper methods that are rather trivial, which are tedious to implement for each user of the library. Conversion methods would also be nice, allowing one type of interval to be converted into another, e.g. for converting a `DateTimeInterval` into a `ZoneDateTimeInterval`/`Interval`.

Are there any plans to support intervals in general in the future? E.g. with more interval data types, helper methods or even data structures. This could either be done within NodaTime, or as a separate project containing the existing interval types.

**Comment by jskeet** [comment]
Each potential interval type would really need to justify its own existence. Not everything makes sense - for example, an interval from one ZonedDateTime to one in a different time zone isn't terribly useful, I suspect. I really think Interval and DateInterval will cover most cases. TimeInterval is a possibility, but with no plans right now - I'd want to see various use cases and a design that handles them nicely.

Interval relation methods *could* be useful - and pull requests are welcome, of course. It becomes slightly tricky when normal Interval values may or may not have a start/end, and DateIntervals can be exclusive or inclusive of end dates.

**Comment by lundmikkel** [comment]
Yeah, ZonedDateTimeInterval might be a bad example (the DateTimeInterval could be converted directly into an Interval though). I work with scheduling and find that we often juggle pairs of values, which are actually a start and end of an interval. The values are given in local times by the user, but converted into global times. Having an interval from the beginning to the end would make it a lot easier to work with.

Relations are not really difficult to make. I have previously made an extensive interval library (with generic intervals), which had plenty of [extension methods](https://github.com/lundmikkel/C5Intervals/blob/master/C5.Intervals/Extensions/IntervalExtensions.cs) which worked with any combination of endpoint inclusions.

**Comment by danygarss** [comment]
Would a ZonedInterval make sense? It would be the combination of an Interval and a DateTimeZone. It could be initialized with those 2 values, or with two ZonedDateTime values, start and end (enforcing that the zone is the same). It would be essentialy the same as an Interval, but with time zone aware, thus Start and End properties would return ZonedDateTime. I can imagine a .InZone(DateTimeZone zone) method in Interval struct, and a .ToInterval() method in this supposed ZonedInterval, to have the same semantics as the conversions between Instant and ZonedDateTime. Not sure if this makes sense, or would be useful to a lot of people. For me it would certainly be, as I usually have to work with intervals of time in a specific time zone.

**Comment by jskeet** [comment]
@danygarss: I *suspect* it wouldn't be very widely used - I suggest you just compose an `Interval` and a `DateTimeZone` in your own type for now - I'm nervous about having *too* many interval-like types.

**Comment by danygarss** [comment]
@jskeet I understand your concerns, this can be easily achieved by anyone who needs it, and maybe it creates noise and aditional stuff to learn for those who don't (also it's aditional stuff to document, test, mantain...)

**Comment by jskeet** [comment]
I think if we want to do anything, we'll want to wait for [C# range support](https://github.com/dotnet/roslyn/blob/features/range/docs/features/range.md).

**Comment by lundmikkel** [comment]
@jskeet I've actually implemented both 'DateTimeInterval` and `TimeInterval` and am currently working on `AnnualDateInterval` since we needed all of them at the company where I work. I think having them has been a great asset to our business and it wasn't an option to keep working without them.

Most of the logic is globally applicable and the code could largely be ported to the main NodaTime library, if that had any interest.

**Comment by Rody66** [comment]
I've also run into this problem and solved it creating a new classes LocalInterval (interval of LocalDateTime) and TimeInterval (interval of LocalTime). Logic for both these classes are based on NodaTime's Interval class. 
Here is the project: https://github.com/Rody66/nodatime.intervals
@jskeet May you look this? And thanks a lot for your great work!

**Comment by jskeet** [comment]
@Rody66: Please create a pull request

**Comment by Rody66** [comment]
> 
> 
> @Rody66: Please create a pull request, assuming you're suggesting new code for within the NodaTime project.
> 
> (I would suggest that there's more complexity than you're already handling - things like different calendar system, and intervals from "10pm - 2am" for example.)

Thanks, I will create a PR. Is it necessary to implement `IXmlSerializable` to new struct?
And about your example with time of the next day - I understand correctly that interval of 23:00/1:00 should return a `Period` of 2 hours, and method `Contains(new LocalTime(0, 0))` should return `true`? If yes, perhaps we should add some Boolean property like `IsEndInNextDay` that will switch to `true` in this case?

**Comment by jskeet** [comment]
I wouldn't worry about serialisation yet. Let's work out whether we definitely want these two types first, and the semantics. (I'm really not sure at the moment.)

**Comment by Rody66** [comment]
PR for TimeInterval created: https://github.com/nodatime/nodatime/pull/1337

**Comment by Rick-Davin** [comment]
> Would a ZonedInterval make sense? It would be the combination of an Interval and a DateTimeZone. It could be initialized with those 2 values, or with two ZonedDateTime values, start and end (enforcing that the zone is the same).
> 
Sorry about chiming in over 3 years too late, but I have played with the notion of a ZonedInterval.  However, I strongly believe that the Start and End (both ZonedDateTime) should be allowed to belong to different DateTimeZone.  Think of air travel.  I depart New York City and fly to London.

In my playing around, I have 2 constructors.  One accepts the Start and End, again both ZonedDateTime.  The other accepts Interval but assumes UTC for both Start and End.  i.e. 

public ZonedInterval(Interval interval) : this(interval.Start.InUtc(), interval.End.InUtc()) { }

A ZonedInterval would have many practical uses for shift work at chemical plants, refineries, hospitals, law enforcement - essentially any where continuous coverage is needed.  Or it could be used for reservation systems, again like air travel or car/truck rentals.

**Comment by jskeet** [comment]
> Sorry about chiming in over 3 years too late, but I have played with the notion of a ZonedInterval. However, I strongly believe that the Start and End (both ZonedDateTime) should be allowed to belong to different DateTimeZone. Think of air travel. I depart New York City and fly to London.

What this says to me is that there isn't a single sufficiently clear use case for `ZonedInterval`. Your air travel case is a good example where the start and end time zones should be different - all the shift work examples would be better in a single time zone, and the flexibility of having different time zones would make those use cases more complex.

I think it's best for now for Noda Time to provide the primitives, and let users create their own interval types for their specific use cases. I definitely prefer to err on the side of caution when it comes to expanding the API. Note that ZonedInterval is definitely more complex than TimeInterval, and even that isn't merged yet because there are multiple use cases that would suggest different APIs.

**Comment by chrbkr** [comment]
Is there a specific reason why Interval is allowed to be open, while DateInterval isn't? For my use cases, the other way around would make a lot more sense: Appointments (which include time) are never open. However, validity periods of all sorts (employment, prices, tariffs etc.) are initially open in most cases, and might be closed later on.

**Comment by jskeet** [comment]
@chrbkr: I wouldn't recommend using Interval for an appointment, as that's more of a "human time" than "machine time" concept. (I'd potentially use a pair of ZonedDateTimes, making them nullable for openness.) As for why things are the way they are - nothing specific; that's just what I felt at the time met the most use cases that I'd come across. (Being able to split "the whole of time" into intervals is definitely useful, which is why they're lower-bound-inclusive, upper-bound-exclusive, for example... whereas DateIntervals are inclusive of both bounds as that's more useful for things like "my vacation period".)

If you want a potentially-open version of DateInterval, it shouldn't be hard - it's very much a "secondary" type, hardly used elsewhere in the code, unlike Interval which is pretty tightly integrated within everything else. (There isn't even text formatting support for DateInterval.)

**Comment by chrbkr** [comment]
@jskeet Thanks for your response. So Interval is mostly for "technical" intervals? I wanted to use this with Postgres + Dapper. Postgres does have [interval](https://www.postgresql.org/docs/current/datatype-datetime.html#DATATYPE-INTERVAL-INPUT), but I guess that would be what Duration is in NodaTime? What I really want to use is [ranges](https://www.postgresql.org/docs/current/rangetypes.html) (tsrange/tstzrange/daterange). So I guess I should write my own classes for that. Or just keep using separate properties.

**Comment by jskeet** [comment]
I'd start by considering the types for each end of the range - Interval is bounded by Instants, whereas for values that originated with humans, I'd usually start with LocalDateTime or ZonedDateTime (assuming both date and time are needed).

I won't offer a view on Postgres as it's easy to be wrong on this sort of thing without being an expert :) (At a quick look it does sound like a Postgres Interval is what I'd call a duration, but it's very very possible that I've missed something.)
